File: stack_queues/456_132_pattern__.py
class Solution:
    def find132pattern(self, nums: List[int]) -> bool:
        stack, last = [], float('-inf')

        for num in reversed(nums):
            if num < last:
                return True
            while stack and stack[-1] < num:
                last = stack.pop()
            stack.append(num)
        return False
        

File: monotonic_stack/456_132_pattern.java
class Solution {
    public boolean find132pattern(int[] nums) {

        Stack<Integer> stack = new Stack<>();
        int[] minimums = new int[nums.length];

        for (int i = 0; i < nums.length; i++) {
            if (i == 0) {
                minimums[i] = 0;
            }
            else if (nums[i] < nums[minimums[i-1]]) {
                minimums[i] = i;
            } else {
                minimums[i] = minimums[i-1];
            }
        }

        for (int i = 0; i < nums.length; i++) {
            while (!stack.empty() && nums[stack.peek()] <= nums[i]) {
                stack.pop();
            }

            if (!stack.empty()) {
                if (nums[minimums[stack.peek()]] < nums[i]) return true;
            }
            stack.push(i);
        }
        return false;
    }
}